module/zfs/abd.c

```diff
@@ -412,6 +432,7 @@ void
 abd_gang_add(abd_t *pabd, abd_t *cabd, boolean_t free_on_free)
 {
 	ASSERT(abd_is_gang(pabd));
+	ASSERT(!(cabd->abd_flags & ABD_FLAG_STACK));
```

Comment:
Well, it would have the same limitations if it's on the stack, regardless of how it's initialised. So no ganging, no moving off thread, etc.

We could expose `abd_init_struct` (it's currently static) but that's still a function call, and initialises a list node and mutex that won't be used. Could move it to the header and inline, but same thing. And then it's slightly more annoying to use, without the assignment.

None of these are very strong arguments, but it seemed like avoiding the allocation was worth it. Maybe it's better to keep it looking the same for the reader though?

Reply:
Why do you keep saying "no ganging", and why no "moving off the thread"?  Sure thread can not get out of the function scope until the ABD is fully processed, but anything else?

Reply:
Ok, to be fair, there's nothing stopping an ABD on the stack being included in a gang ABD (well, implementation right now doesn't initialise the mutex and list node, but lets say it did).

I guess I was more thinking that we really only do these kind of things when we're dispatching some kind of async task (IO), so if someone is doing that, they've likely screwed up, and debugging it is likely to be _extremely_ difficult. That's why the flag and the asserts are there, to try to make it clear that you should probably be using a fully allocated ABD at this point.

(I did even consider a current thread check or stack pointer check, but decided that was far too invasive).

Honestly, I'm not committed to this, it just seemed like an easy argument would be to now require an allocation and deallocation just to wrap a buffer. In some future "everything" might be ABD and wrapping won't be necessary at all.

Is your objection to the concept generally, to the limits I've placed on it, or to my explanation of it?

Reply:
I don't like introduction of one more ABD type with special characteristics.

Reply:
Do you like it less or more than needing an allocation to wrap a buffer just to deallocate it immediately after? That's really the whole question. That is:

```c
abd_t *abd = abd_get_from_buf(buf, len);
do_the_abd_thing(abd);
abd_free(abd);
```

This is what I had right up until near the end of this work, and I can certainly live with going back to it.

Reply:
I from the beginning proposed something like this:
```
        abd_t abd;
        abd_get_from_buf_struct(&abd, buf, len);
        do_the_abd_thing(&abd);
        abd_free(&abd);
```
I.e. no allocation/free, but otherwise fully functional ABD of already existing type.

Reply:
Ok, last creates `abd_get_from_buf_struct()` and uses it in place of `ABD_STACK_INIT()`. The bulk of the diff is of this shape:

```diff
-       abd_t sabd = ABD_STACK_INIT(src, s_len);
-       abd_t dabd = ABD_STACK_INIT(dst, d_len);
+       abd_t sabd, dabd;
+       abd_get_from_buf_struct(&sabd, src, s_len);
+       abd_get_from_buf_struct(&dabd, dst, d_len);
        err = zio_decompress_data(cpfunc, &sabd, &dabd, s_len, d_len, NULL);
+       abd_free(&dabd);
+       abd_free(&sabd);
```

I personally do not enjoy using this anywhere near as much. I will however agree that introducing an ABD variant that behaves subtly different in some circumstances is probably not the right thing for this PR. I'll hol don to the patch for now and maybe it'll see light of day in some future ABD API upgrade.

Thanks!